olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/util/SubprocessConnection.java

```diff
@@ -192,112 +292,230 @@ public void setTimeout(int time, TimeUnit unit) {
      * any text that is returned up until an empty line is printed. A string containing
      * the returned text, suitable for being wrapped in a StringReader, is returned.
      *
+     * This method respects the read timeout that is set with {@link #setReadTimeout(int, TimeUnit)}.
+     * If no text is read from the subprocess while waiting for input after the readTimeout
+     * elapses, a TimeoutException will be thrown. Each time the subprocess generates a
+     * new line, the timer is reset.
+     *
      * @param command the text to send to the subprocess
      * @return all text returned from the subprocess
+     * @throws IOException if an error occurred reading from the subprocess
+     * @throws TimeoutException if reading from the subprocess took too long
      */
-    public String run(String command) throws IOException {
-        StringBuilder results = new StringBuilder();
-        ensureRunning();
-        synchronized (process) {
+    public synchronized String run(String command) throws IOException, TimeoutException {
+        //
+        // This method is synchronized to limit the threads that can wait on the processLock.
+        // If a process is killed, we want the process reaper to grab the lock before anybody
+        // else can, so keeping threads waiting above this method helps make that possible.
```

Comment:
I guess I don't understand how this ensures that the subprocess reaper gets the lock. Isn't that caused by the timeout?